core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1289,24 +1308,50 @@ public void run() {
             channelLogger.log(ChannelLogLevel.INFO, "Address resolved: {0}", servers);
             haveBackends = true;
           }
-          final Map<String, ?> serviceConfig =
+
+          // Assuming no error in config resolution for now.
+          Map<String, ?> serviceConfig =
               config.get(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG);
-          if (serviceConfig != null && !serviceConfig.equals(lastServiceConfig)) {
-            channelLogger.log(ChannelLogLevel.INFO, "Service config changed");
+          Attributes effectiveConfig = config;
+          if (!lookUpServiceConfig || serviceConfig == null) {
```

Comment:
Maybe simplify the conditions and make it flow better:

```java
Map<String, ?> effectiveConfig = null;

// Try to use config if returned from name resolver and allowed
if (serviceConfig != null) {
  if (lookUpServiceConfig) {
    effectiveConfig = serviceConfig;
  } else {
    channelLogger.log(ChannelLogLevel.INFO, "Service config from name resolver discarded by channel settings");
  }
}

// Otherwise, try to use the default config if available
if (effectiveConfig == null && defaultServiceConfig != null) {
  channelLogger.log(ChannelLogLevel.INFO, "Using default service config");
  effectiveConfig = defaultServiceConfig;
}

if (effectiveConfig != lastServiceConfig) {
  channelLogger.log(ChannelLogLevel.INFO,
      "Service config changed" + (effectiveConfig == null ? " to null" : ""));
  lastServiceConfig = effectiveConfig;
}

config = config.toBuilder()
    .set(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG, effectiveConfig)
    .build();

waitingForServiceConfig = false;

```

Reply:
Done.